### PR TITLE
gh-697 Ensure an AppRegistration exists before deleting it

### PR DESCRIPTION
**Pull request by ghillert**
* Add several tests for the `/apps` REST endpoint
* Refactor `AppAlreadyRegisteredException` into a separate class
* Add proper `ExceptionHandler` for `AppAlreadyRegisteredException` in `RestControllerAdvice`
  - Returning HTTP status `409` (Conflict)
* Add proper `ExceptionHandler` for `NoSuchAppRegistrationException` in `RestControllerAdvice`
  - Returning HTTP status `404` (Not Found)

Resolves spring-cloud/spring-cloud-dataflow#697

### CODE REVIEW

File: spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/AppRegistry.java [hunk 1]
```diff
@@ -81,8 +83,22 @@ public AppRegistration save(String name, ApplicationType type, URI uri) {
...
+	 * Deletes an {@link AppRegistration}. If the {@link AppRegistration} does not
+	 * exist, a {@link NoSuchAppRegistrationException} will be thrown.
+	 *
+	 * @param name Name of the AppRegistration to delete
+	 * @param type Type of the AppRegistration to delete
+	 */
 	public void delete(String name, ApplicationType type) {
-		this.uriRegistry.unregister(key(name, type));
+		final AppRegistration appRegistrationToDelete = this.find(name, type);
+
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/AppRegistry.java]: extra line can be avoided.

File: spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/AppRegistry.java [hunk 2]
```diff
@@ -81,8 +83,22 @@ public AppRegistration save(String name, ApplicationType type, URI uri) {
...
+		if (appRegistrationToDelete != null) {
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/AppRegistry.java]: Given `appRegistrationToDelete` is used only in this line, can we move `find` here directly and perform null check on return value?

File: spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/support/package-info.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+/**
+ * Contains {@link org.springframework.cloud.dataflow.registry.AppRegistration}
+ * support classes.
+ */
+package org.springframework.cloud.dataflow.registry.support;
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-registry/src/main/java/org/springframework/cloud/dataflow/registry/support/package-info.java]: new line.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/AppAlreadyRegisteredException.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+	@Override
+	public String getMessage() {
+		return String.format("The '%s:%s' application is already registered as %s", previous.getType(), previous.getName(), previous.getUri());
+	}
+
+	public AppRegistration getPrevious() {
+		return previous;
+	}
+}
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/AppAlreadyRegisteredException.java]: newline.

**Comment by ghillert** [comment]
@ilayaperumalg Addressed the raised issues

**Comment by ilayaperumalg** [comment]
LGTM, squashed, rebased and merged as 3f867f735473835cacdd3c038127039d4935ac37

**Pull request closed without merging**